### PR TITLE
S3 Persist

### PR DESCRIPTION
**Pull request by m3nadav**

### CODE REVIEW

**Comment by shaharke** [comment]
@m3nadav don't merge to master

File: lib/s3.persist.stream.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      this._file = s3Stream.upload({
+        "Bucket": "bp-objects-cache",
+        "Key": this._filename
+      });
+
+      this._file.on('error', function(e) {
+        self._writeError = e;
+      })
+
+      /*
```
Comment (shaharke) [inline lib/s3.persist.stream.js]: Remove

File: lib/s3.persist.stream.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    } catch (e) {
+      this._writeError = e;
+    }
+
+  }
+
+  self.push(chunk);
+
+  if (!this._writeError) {
+    return this._file.write(chunk, enc, done);
```
Comment (shaharke) [inline lib/s3.persist.stream.js]: This means that the backpressure is a function of the S3 upload rate and not the next stream in the metadata.

File: lib/s3.persist.stream.js (hunk 1 above)
Comment (m3nadav) [inline lib/s3.persist.stream.js]: this pr is not ready yet, but yeah this will be removed

File: lib/s3.persist.stream.js (hunk 2 above)
Comment (m3nadav) [inline lib/s3.persist.stream.js]: not exectly, the s3 upload stream package is buffering data before uploading so most of the chunks are only buffered.
When the package uploads the data to s3 it does it async

**Comment by m3nadav** [comment]
@shaharke @dshush 
now its time to review 

same thing about rotation as i said in metadata_api

File: README.md [hunk 3]
```diff
@@ -35,6 +35,24 @@ __options:__
...
+```javascript
+var S3PersistStream = require('bp-streams').S3PersistStream;
+var stream = new S3PersistStream(options);
+```
+
+__options:__
+
+| Name          | Description           | Default  |
+| :------------ |:-------------| :-----:|
+| metadata_version | filename | - |
```
Comment (dshush) [inline README.md]: I would name the options here for what they would be use and not what values we will put in them, metadata_version is true for the specific case of metadata persist - but we can use the S3PersistStream also for other cases. The same about the stream_id.

File: lib/s3.persist.stream.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
+var AWS = require('aws-sdk');
+var _ = require('underscore');
+
+var stream = require('stream');
+var util = require('util');
+
+var rotateHistoryCount = 3;
```
Comment (dshush) [inline lib/s3.persist.stream.js]: I think this should also be a configuration received from outside - we might want to have different rotation for every use case

File: tests/s3.persist.stream.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+var S3PersistStream = require('..').S3PersistStream;
```
Comment (dshush) [inline tests/s3.persist.stream.js]: Forgot to put .spec in the end of the file

File: lib/s3.persist.stream.js (hunk 4 above)
Comment (m3nadav) [inline lib/s3.persist.stream.js]: i think ill leave this comment unresolved untill we talk with oschribe  about how to rotate

File: tests/s3.persist.stream.js (hunk 5 above)
Comment (m3nadav) [inline tests/s3.persist.stream.js]: ohh i copied older file i guess before the conventions were made

File: lib/s3.persist.stream.js (hunk 4 above)
Comment (shaharke) [inline lib/s3.persist.stream.js]: I agree. I actually have my own reservations on who should be responsible for the rotation.

File: lib/s3.persist.stream.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+  if (!options.stream_id) {
+    throw new Error('Stream ID must be specified for S3 persist stream');
+  }
+
+  if (!options.bucket) {
+    throw new Error('Bucket must be specified for S3 persist stream');
+  }
+
+  this._metadata_version = options.metadata_version;
```
Comment (shaharke) [inline lib/s3.persist.stream.js]: See @dshush comment above

File: lib/s3.persist.stream.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+util.inherits(S3PersistStream, PassThrough);
+
+S3PersistStream.prototype._write = function (chunk, enc, done) {
+  var self = this;
+
+  if (!this._writeError && this._uploadStream == null) {
+    try {
+      this._uploadStream = self.s3Stream.upload(this._uploadParams);
+      this._uploadStream.maxPartSize(10485760);
```
Comment (shaharke) [inline lib/s3.persist.stream.js]: Should be configurable

File: lib/s3.persist.stream.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+      this._writeError = e;
+      this.emit('error', this._writeError);
+    }
+
+  }
+
+  self.push(chunk);
+
+  if (!this._writeError) {
+    return this._uploadStream.write(chunk, enc, done);
```
Comment (shaharke) [inline lib/s3.persist.stream.js]: My comment from before (haven't seen your response): aren't we back-pressuring on the wrong stream?

File: package.json [hunk 9]
```diff
@@ -22,8 +22,14 @@
   "devDependencies": {
```
Comment (shaharke) [inline package.json:22]: Use fixed version for all dependencies. You can actually configure npm to use fixed version by default: in ~/.npmrc add `save-exact=true`

File: tests/s3.persist.stream.js (hunk 5 above)
Comment (shaharke) [inline tests/s3.persist.stream.js]: Can you also fix the older files? :)

File: tests/s3.persist.stream.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+var AWS = require('aws-sdk');
+var s3 = new AWS.S3();
```
Comment (shaharke) [inline tests/s3.persist.stream.js]: ah... use mock instead? what credentials are you using for the tests?

**Comment by shaharke** [comment]
don't merge to master

**Comment by shaharke** [comment]
Also - notice that this is a public repo. Don't use anything that might relate to BigPanda in particular.

File: lib/s3.persist.stream.js (hunk 8 above)
Comment (m3nadav) [inline lib/s3.persist.stream.js]: lets talk about it

**Pull request closed without merging**